File: client/src/utils/composeResume.ts
import type { Profile, ResumeProfile } from '../api/profile';

/**
 * Output of composeResume. `text` is the plain string the copy button puts on the clipboard;
 * `boldLines` is the subset of `text` lines (by verbatim string) the viewer should render as
 * bold/larger — currently the experience header lines we synthesize from
 * `[Subtitle N]` + profile.experiences[N-1]. Plain text stays paste-ready.
 */
export type ComposedResume = {
  text: string;
  boldLines: string[];
};

function nameForProfile(p: ResumeProfile | Profile | null | undefined): string {
  if (!p) return '';
  if (p.displayName) return p.displayName;
  return 'nickname' in p && p.nickname ? (p.nickname as string) : '';
}

function formatExperienceHeader(
  role: string,
  exp: ResumeProfile['experiences'][number] | undefined
): string {
  if (!exp) return role.trim();
  const range =
    exp.startYear && exp.endYear
      ? `${exp.startYear} - ${exp.endYear}`
      : exp.startYear
        ? `${exp.startYear} -`
        : exp.endYear
          ? `- ${exp.endYear}`
          : '';
  return [role, exp.company, exp.location, range]
    .map((s) => (s || '').trim())
    .filter(Boolean)
    .join(' · ');
}

/**
 * Walk the bid body line-by-line and apply substitution rules:
 *   `[Title]`/`[Title]:` / `[FolderName]`/`[FolderName]:` / bare `Edit`
 *     → strip the line entirely.
 *   `[Subtitle N]`/`[Subtitle N]:`
 *     → strip the marker AND consume the next non-empty content line as roleN[N].
 *   `[Experience N]`/`[Experience N]:`
 *     → replace with a header `roleN[N] · company · location · 2022 - 2024`, drawn from
 *       profile.experiences[N-1]. Falls back gracefully when a piece is missing. Header
 *       strings are recorded in `boldLines` so the viewer can render them bigger.
 *   First non-empty line that matches the profile name (case-insensitive)
 *     → strip (legacy template variant where GPT prepended the user's name).
 *   `[Summary]`/`[Summary]:` / `[Skills]`/`[Skills]:` lines pass through literal.
 */
function applyPlaceholders(
  body: string,
  experiences: ResumeProfile['experiences'],
  profileName: string
): { lines: string[]; boldSet: Set<string> } {
  const inLines = body.split('\n');
  const roleByIndex = new Map<number, string>();
  const stripIndex = new Set<number>();
  /** Pre-pass 1: capture each [Subtitle N] role from the next non-empty line, mark both stripped. */
  for (let i = 0; i < inLines.length; i++) {
    const m = inLines[i].trim().match(/^\[subtitle\s+(\d+)\]:?\s*$/i);
    if (!m) continue;
    stripIndex.add(i);
    for (let j = i + 1; j < inLines.length; j++) {
      if (inLines[j].trim() === '') continue;
      roleByIndex.set(Number(m[1]), inLines[j].trim());
      stripIndex.add(j);
      break;
    }
  }
  /** Pre-pass 2: drop a leading "JOSHUA"-style line that duplicates the profile display name. */
  if (profileName) {
    const pname = profileName.trim().toLowerCase();
    for (let i = 0; i < inLines.length; i++) {
      const t = inLines[i].trim();
      if (!t) continue;
      if (t.toLowerCase() === pname) stripIndex.add(i);
      break;
    }
  }

  const out: string[] = [];
  const boldSet = new Set<string>();
  for (let i = 0; i < inLines.length; i++) {
    if (stripIndex.has(i)) continue;
    const raw = inLines[i];
    const trimmed = raw.trim();

    if (/^\[title\]:?\s*$/i.test(trimmed)) continue;
    if (/^\[foldername\]:?\s*$/i.test(trimmed)) continue;
    if (trimmed.toLowerCase() === 'edit') continue;

    const expMatch = trimmed.match(/^\[experience\s+(\d+)\]:?\s*$/i);
    if (expMatch) {
      const idx = Number(expMatch[1]) - 1;
      const exp = idx >= 0 ? experiences[idx] : undefined;
      const role = roleByIndex.get(idx + 1) ?? '';
      const header = formatExperienceHeader(role, exp);
      if (header) {
        out.push(header);
        boldSet.add(header);
      }
      continue;
    }

    out.push(raw);
  }
  return { lines: out, boldSet };
}

/**
 * Build a resume-shaped string by combining the row owner's profile header with the per-bid
 * resume body (the `gptResumeContent` saved by the Bid Assistant). Returns null when there's
 * nothing to show. `profile` is the per-group profile of the bid owner (viewer in self view,
 * bidder in caller view); falls back to the user-level profile when the group profile hasn't
 * been seeded yet.
 */
export function composeResume(
  profile: ResumeProfile | Profile | null | undefined,
  body: string
): ComposedResume | null {
  const trimmedBody = (body || '').trim();
  if (!profile && !trimmedBody) return null;

  /** Only ResumeProfile has experiences; fall back to [] when given the user-level Profile. */
  const experiences =
    profile && 'experiences' in profile && Array.isArray(profile.experiences)
      ? profile.experiences
      : [];

  const placeholderResult = trimmedBody
    ? applyPlaceholders(trimmedBody, experiences, nameForProfile(profile))
    : { lines: [], boldSet: new Set<string>() };
  const processedBody = placeholderResult.lines
    .join('\n')
    .replace(/\n{3,}/g, '\n\n')
    .trim();

  const lines: string[] = [];

  if (profile) {
    const name = nameForProfile(profile);
    if (name) lines.push(name.toUpperCase());
    if (profile.headline) lines.push(profile.headline);

    const contactBits: string[] = [];
    if (profile.location) contactBits.push(profile.location);
    if (profile.personalEmail) contactBits.push(profile.personalEmail);
    if (profile.phone) contactBits.push(profile.phone);
    if (contactBits.length > 0) lines.push(contactBits.join(' | '));
    if (profile.linkedinUrl) lines.push(profile.linkedinUrl);
  }

  if (processedBody) {
    if (lines.length > 0) lines.push('');
    lines.push(processedBody);
  }

  if (profile && profile.education.length > 0) {
    lines.push('');
    lines.push('Education');
    for (const e of profile.education) {
      const range =
        e.startYear && e.endYear
          ? `${e.startYear} - ${e.endYear}`
          : e.startYear
            ? `${e.startYear} -`
            : e.endYear
              ? `- ${e.endYear}`
              : '';
      const parts = [e.degree, e.school, e.location, range].filter(Boolean);
      if (parts.length > 0) lines.push(parts.join(' · '));
    }
  }

  if (profile && profile.certifications.length > 0) {
    lines.push('');
    lines.push('Certifications');
    for (const c of profile.certifications) {
      const parts = [c.name, c.issuer, c.year != null ? String(c.year) : ''].filter(Boolean);
      if (parts.length > 0) lines.push(parts.join(' · '));
    }
  }

  const text = lines.join('\n').trim();
  if (!text) return null;
  return { text, boldLines: [...placeholderResult.boldSet] };
}

File: desktop/DevStrider.Desktop/Services/ComposeResume.cs
using System.Text.RegularExpressions;
using DevStrider.Desktop.Models;

namespace DevStrider.Desktop.Services;

/// <summary>
/// Port of <c>client/src/utils/composeResume.ts</c>: build a paste-ready resume by combining the
/// profile header with a per-bid body that contains placeholder markers from GPT. The viewer
/// gets back both plain text (for copy) and a list of lines to render in bold/larger weight
/// (the synthesized experience headers).
/// </summary>
public static class ComposeResume
{
    public record Composed(string Text, IReadOnlyList<string> BoldLines);

    public static Composed? Build(UserProfile? profile, string body)
    {
        var trimmedBody = (body ?? "").Trim();
        if (profile == null && trimmedBody.Length == 0) return null;

        var experiences = profile?.Experiences ?? new List<Experience>();
        var processed = trimmedBody.Length > 0
            ? ApplyPlaceholders(trimmedBody, experiences, profile?.DisplayName ?? "")
            : (new List<string>(), new HashSet<string>());

        var bodyText = string.Join("\n", processed.Lines).Trim();
        // collapse 3+ blank lines (stripped placeholder leftovers).
        bodyText = Regex.Replace(bodyText, "\n{3,}", "\n\n");

        var lines = new List<string>();
        if (profile != null)
        {
            var name = string.IsNullOrEmpty(profile.DisplayName) ? "" : profile.DisplayName.ToUpperInvariant();
            if (name.Length > 0) lines.Add(name);
            if (!string.IsNullOrEmpty(profile.Headline)) lines.Add(profile.Headline);

            var contact = new List<string>();
            if (!string.IsNullOrEmpty(profile.Location)) contact.Add(profile.Location);
            if (!string.IsNullOrEmpty(profile.PersonalEmail)) contact.Add(profile.PersonalEmail);
            if (!string.IsNullOrEmpty(profile.Phone)) contact.Add(profile.Phone);
            if (contact.Count > 0) lines.Add(string.Join(" | ", contact));
            if (!string.IsNullOrEmpty(profile.LinkedinUrl)) lines.Add(profile.LinkedinUrl);
        }

        if (bodyText.Length > 0)
        {
            if (lines.Count > 0) lines.Add("");
            lines.Add(bodyText);
        }

        if (profile?.Education.Count > 0)
        {
            lines.Add("");
            lines.Add("Education");
            foreach (var e in profile.Education)
            {
                var range =
                    e.StartYear != null && e.EndYear != null ? $"{e.StartYear} - {e.EndYear}" :
                    e.StartYear != null ? $"{e.StartYear} -" :
                    e.EndYear != null ? $"- {e.EndYear}" : "";
                var parts = new[] { e.Degree, e.School, e.Location, range }
                    .Where(s => !string.IsNullOrWhiteSpace(s));
                if (parts.Any()) lines.Add(string.Join(" · ", parts));
            }
        }

        if (profile?.Certifications.Count > 0)
        {
            lines.Add("");
            lines.Add("Certifications");
            foreach (var c in profile.Certifications)
            {
                var parts = new[] { c.Name, c.Issuer, c.Year?.ToString() ?? "" }
                    .Where(s => !string.IsNullOrWhiteSpace(s));
                if (parts.Any()) lines.Add(string.Join(" · ", parts));
            }
        }

        var text = string.Join("\n", lines).Trim();
        if (text.Length == 0) return null;
        return new Composed(text, processed.Bold.ToList());
    }

    private static (List<string> Lines, HashSet<string> Bold) ApplyPlaceholders(
        string body, IReadOnlyList<Experience> experiences, string profileName)
    {
        var inLines = body.Replace("\r\n", "\n").Split('\n');
        var roleByIndex = new Dictionary<int, string>();
        var strip = new HashSet<int>();

        // Pass 1: capture [Subtitle N] roles, mark both lines stripped.
        for (int i = 0; i < inLines.Length; i++)
        {
            var t = inLines[i].Trim();
            var m = Regex.Match(t, @"^\[subtitle\s+(\d+)\]:?\s*$", RegexOptions.IgnoreCase);
            if (!m.Success) continue;
            strip.Add(i);
            for (int j = i + 1; j < inLines.Length; j++)
            {
                if (inLines[j].Trim().Length == 0) continue;
                roleByIndex[int.Parse(m.Groups[1].Value)] = inLines[j].Trim();
                strip.Add(j);
                break;
            }
        }

        // Pass 2: drop a leading line that duplicates the profile display name (legacy template).
        if (!string.IsNullOrEmpty(profileName))
        {
            var pname = profileName.Trim().ToLowerInvariant();
            for (int i = 0; i < inLines.Length; i++)
            {
                var t = inLines[i].Trim();
                if (t.Length == 0) continue;
                if (t.ToLowerInvariant() == pname) strip.Add(i);
                break;
            }
        }

        var outLines = new List<string>();
        var bold = new HashSet<string>();
        for (int i = 0; i < inLines.Length; i++)
        {
            if (strip.Contains(i)) continue;
            var raw = inLines[i];
            var t = raw.Trim();
            if (Regex.IsMatch(t, @"^\[title\]:?\s*$", RegexOptions.IgnoreCase)) continue;
            if (Regex.IsMatch(t, @"^\[foldername\]:?\s*$", RegexOptions.IgnoreCase)) continue;
            if (t.Equals("edit", StringComparison.OrdinalIgnoreCase)) continue;

            var m = Regex.Match(t, @"^\[experience\s+(\d+)\]:?\s*$", RegexOptions.IgnoreCase);
            if (m.Success)
            {
                var n = int.Parse(m.Groups[1].Value);
                var role = roleByIndex.GetValueOrDefault(n, "");
                var exp = n - 1 >= 0 && n - 1 < experiences.Count ? experiences[n - 1] : null;
                var range =
                    exp?.StartYear != null && exp?.EndYear != null ? $"{exp.StartYear} - {exp.EndYear}" :
                    exp?.StartYear != null ? $"{exp.StartYear} -" :
                    exp?.EndYear != null ? $"- {exp.EndYear}" : "";
                var parts = new[] { role, exp?.Company ?? "", exp?.Location ?? "", range }
                    .Select(s => (s ?? "").Trim())
                    .Where(s => s.Length > 0);
                var header = string.Join(" · ", parts);
                if (header.Length > 0)
                {
                    outLines.Add(header);
                    bold.Add(header);
                }
                continue;
            }
            outLines.Add(raw);
        }
        return (outLines, bold);
    }
}
